languages/ruby/exercises/concept/ostruct/.docs/introduction.md

```diff
@@ -0,0 +1,33 @@
+Ruby comes with a Standard Library - a collection of classes for working with things such as dates, json, and networking.
+The also provide some useful functionality for making your code easier to work with.
+
+`OpenStruct` is part of the standard library and allows you to easily create an object from a hash. Rather than having to access using hash keys, `OpenStruct` instead allows us to use methods to access and set values.
+
+When using classes from the Standard Library, or any other library, you need to require that class using the `require` method.
+
+```ruby
+require 'ostruct'
+
+attributes = { name: "Jeremy Walker", age: 21, location: "Nomadic" }
+person = OpenStruct.new(attributes)
+
+person.name
+#=> Jeremy Walker
+
+person.location
+#=> Nomadic
+
+# Update the age
+person.age = 35
+
+# It sets correctly
+person.age
+#=> 21
+```
+
+One bonus advantage of this is that you can take advantage of a shortcut when using block syntax. In situations where a block calls a single method, you can replace the block with `&:` followed by the method name. For example, these two lines are synonymous:
```

Comment:
```suggestion
This allows us to take advantage of a shortcut when using block syntax. In situations where a block calls a single method, you can replace the block with `&:` followed by the method name. For example, these two lines are synonymous:
```

Reply:
Why is this different than defining a method ourselves and gaining the "shortcut" when using block syntax?  How is it better?  So is it the magic of OpenStruct that is giving us that benefit?  Or is it because we aren't using a Hash?  (In other words, show the way, maybe of doing the same thing not using `OpenStruct`. (or `Struct`… why are we going to Standard Library here instead of exploring `Struct`, again?)

Reply:
> Why is this different than defining a method ourselves and gaining the "shortcut" when using block syntax? 

Because we don't have to define a method ourselves?

What would the equivelent implementation of this be with a Struct:
```
    @items = items.map { |item| OpenStruct.new(item) }
```

Reply:
For the "metaprogramming" nature of it?  That is what I am reading from this (I may be off-track though).

If we don't know what our keys are going to be, then we might do this with Struct:

```ruby
inventory = Struct.new(*items[0].keys)
manifest = items.map {|item| inventory.new(*item.values)}
manifest.sum(&:price)
```
My sample `items` list (`Array` of `Hash`es:

```ruby
items = [
  {
    name: 'Rose: yellow',
    price: 1.99,
    description: 'Single yellow rose, long stem'
  },
  {
    name: 'Tulip',
    price: 1.20,
    description: 'Red tulip, freshly picked from our own garden'
  }
]
```

I will admit, I don't use either of them very often.  Last time I looked at the performance and details of using them was likely in 2014, when OpenStruct was slower because at east instantiation of OpenStruct memory was remapped, which made it a fairly expensive thing to use if it was instantiating often.

Reply:
So simply:
```ruby
 @items = items.map { |item| OpenStruct.new(item) }
```

Is just much simpler to get your head around than:

```ruby
inventory = Struct.new(*items[0].keys)
@items = items.map {|item| inventory.new(*item.values)}
```

That's why I'd use OpenStruct :)

Reply:
Well, the exercise is all about `OpenStruct`, so it makes sense, of course, to use it.